planner/core/cache.go

```diff
@@ -147,12 +147,18 @@ func NewPSTMTPlanCacheKey(sessionVars *variable.SessionVars, pstmtID uint32, sch
 type FieldSlice []types.FieldType
 
 // Equal compares FieldSlice with []*types.FieldType
+// Currently this is only used in plan cache to invalidate cache when types of variables are different.
 func (s FieldSlice) Equal(tps []*types.FieldType) bool {
 	if len(s) != len(tps) {
 		return false
 	}
 	for i := range tps {
-		if !s[i].Equal(tps[i]) {
+		// We only use part of logic of `func (ft *FieldType) Equal(other *FieldType)` here because (1) only numeric and
+		// string types will show up here, and (2) we don't need flen and decimal to be matched exactly to use plan cache
```

Comment:
IMHO, this assumption is pretty fragile. Consider this case:

```
MySQL [test]> create table t(a varchar(20), b int, index ia(a(4)));
Query OK, 0 rows affected (0.12 sec)

MySQL [test]> explain select * from t use index(ia) where a = 'xx';
+-------------------------------+---------+-----------+----------------------+---------------------------------------------------+
| id                            | estRows | task      | access object        | operator info                                     |
+-------------------------------+---------+-----------+----------------------+---------------------------------------------------+
| IndexLookUp_8                 | 10.00   | root      |                      |                                                   |
| ├─IndexRangeScan_5(Build)     | 10.00   | cop[tikv] | table:t, index:ia(a) | range:["xx","xx"], keep order:false, stats:pseudo |
| └─Selection_7(Probe)          | 10.00   | cop[tikv] |                      | eq(test.t.a, "xx")                                |
|   └─TableRowIDScan_6          | 10.00   | cop[tikv] | table:t              | keep order:false, stats:pseudo                    |
+-------------------------------+---------+-----------+----------------------+---------------------------------------------------+
4 rows in set (0.00 sec)
```

If we apply an optimization in the future to check the length of the constant string and decide if we need to keep the `Selection_7`, then this cached plan may return wrong results if we reuse it.

For this case, we can check `curr_flen <= cached_flen` to avoid potential bugs, but I am not sure if there are other possibilities that our plan may be dependent on the `flen` / `decimal` of the constants.

Reply:
it's better to use another function to perform this special `equal` validation for plan cache.

Reply:
After discussion, we decide to leave behind the prefix index case mentioned above for now.
Because in `getPhysicalPlan`, we don't know exactly how will a user variable be used. It might be inside a filter expression or part of a range.
If now we want to prevent the case, we can expect that we'll need to implement something like "find if this user variable is used on some indexes in this plan", which is complicated.
It will be easier just not to cache that plan when we implemented this optimization someday.
Also, I added some comments to clarify this.